base/loading.jl

```diff
@@ -1884,8 +1884,17 @@ function _require(pkg::PkgId, env=nothing)
                     @goto load_from_cache
                 end
                 # spawn off a new incremental pre-compile task for recursive `require` calls
-                cachefile = compilecache(pkg, path)
-                if isa(cachefile, Exception)
+                cachefile_or_module = maybe_cachefile_lock(pkg, path) do
```

Comment:
So in https://github.com/JuliaLang/Pkg.jl/issues/3458 we found it to be possible for a fork-bomb to occur.
That makes me wonder how this will handle accidental recursion. Will it now just hang since the parent obtains the pid-lock, then shells out, and the child hangs?

Reply:
We could maybe even try to detect this specific deadlock scenario by calling `getppid` https://linux.die.net/man/3/getppid

The other question is how robust is this against cancellation, what happens on a `killall -9 julia`